### ISSUE TITLE
Timeout does not work if native code is running at the time

### ISSUE THREAD

**Issue by jnm2** [nunit/nunit#1996]
If you have native code running, Timeout has no effect. I assume this is because it relies on Thread.Abort.

Is there anything that could be done to improve this? A native way to nuke it if Thread.Abort does nothing?

**Comment by ChrisMaddock** [comment]
https://github.com/nunit/nunit/issues/1363 is probably relevant here. Just springs to mind as I came across it only a couple of days ago with regards to making Timeout available in the netstandard build, where System.Threading isn't available. I was previously considering a separate implementation using Tasks (which wouldn't be available in <net-4ish) - but sounds like a non-thread-based implementation of TimeoutAttribute could solve all three problems.

@CharliePoole - what thoughts did you have on the above issue, as its assigned to you?

**Comment by CharliePoole** [comment]
The intent of #1363 is to cancel the entire worker thread when there is a timeout, thereby eliminating the need for a "side thread" to run those tests with timeout. To do that, we need a way to add and remove workers smoothly while the tests are running. I think that's a general improvement to the dispatcher that we could use and I'd like to work on it.

Ideally, it would be great if we could deal with timeout in a different way. In that case, I'd re-title #1363 and just do the enabling bit, not the timeout.

Even if your implementation using tasks were only available in .NET 4+ and NetStandard, I think it would be a great improvement, so I say go for it!

**Comment by jnm2** [comment]
Timeouts also have no effect if the execution is inside a call to `Application.Run()`. My original question was kind of specialized because I was calling MsgWaitForMultipleObjectsEx, but this is much more general of a problem.

**Comment by CharliePoole** [comment]
Are we just doing a simple Thread.Abort? We used to have code that did a bit more and took care of some of these cases.

**Comment by jnm2** [comment]
I would have expected Thread.Abort to work against the WinForms message pump, since I would have expected all waits and joins in the entire framework to be hardened for thread abort.

Apparently it just calls [either MsgWaitForMultipleObjectsEx](https://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/Application.cs,2182) or [WaitMessage](https://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/Application.cs,2195) without waiting on any kind of 'thread abort' handle.

The solution seems to be kind of easy perhaps: just send a thread message and the native to managed transition will abort the thread.

**Comment by jnm2** [comment]
For a workaround I'll try writing code that waits till the test timeout and then starts sending messages to the thread.

**Comment by CharliePoole** [comment]
If you find any new wrinkles that work, consider adding them to ThreadUtility.Kill, which is used to terminate threads on timeout.

**Comment by jnm2** [comment]
The proof of concept is working. I just need to replace my message-sending mechanism (`threadSynchronizationContext.Post(_ => { }, null`) with one that sends a native message and doesn't need a SynchronizationContext. `NativeWindow` requires a Windows.Forms reference, so I'll probably just pinvoke to post the message.

**Comment by jnm2** [comment]
In the test:

```c#
using (NativeTimeoutHelper.Run())
    Application.Run();
```

Initial draft:
```c#
public static class NativeTimeoutHelper
{
    public static IDisposable Run()
    {
        if (Debugger.IsAttached) return null;

        var timeout = TestContext.CurrentContext.Test.Properties["Timeout"].Cast<int?>().Min();
        if (timeout == null) return null;

        var finishedRunning = new CancellationTokenSource();
        var thread = Thread.CurrentThread;
        var syncContext = SynchronizationContext.Current;

        Task.Delay(timeout.Value, finishedRunning.Token).ContinueWith(async delayTask => // Continue on threadpool thread
        {
            while (true)
            {
                switch (thread.ThreadState)
                {
                    case ThreadState.Aborted:
                        return;
                    case ThreadState.AbortRequested:
                        syncContext.Post(_ => { }, null);
                        break;
                }
                await Task.Delay(100, finishedRunning.Token);
            }
        }, finishedRunning.Token);

        return On.Dispose(finishedRunning.Cancel);
    }
}
```

**Comment by jnm2** [comment]
Definition of [On.Dispose](https://gist.github.com/jnm2/a3219243836f20e5eb4f8ef777ae4e1c)

**Comment by jnm2** [comment]
It retries every 100ms until the thread is aborted. It's usually ~150 to 200 ms later than the timeout time but that's not important to me.

**Comment by jnm2** [comment]
Not having any luck with `PostThreadMessage`, so I'll have to enumerate windows and send one of them a message.

**Comment by jnm2** [comment]
`PostThreadMessage` `WM_CLOSE` is working well. This will work even without getting a SynchronizationContext which is good because they change.

**Comment by jnm2** [comment]
This solves my original problem and the Application.Run problem.

Here it is: https://gist.github.com/jnm2/0dbc9e5f031d4f662fe591dca818e1d1
Definition of `On.Dispose`: https://gist.github.com/jnm2/a3219243836f20e5eb4f8ef777ae4e1c

---

Next problem: Some tests are stuck in unmanaged code not at a message pool but inside `System.Globalization.CompareInfo.Compare(string, string, CompareOptions)`. Nothing can help that situation I guess except TerminateThread.

**Comment by jnm2** [comment]
Next problem: Some tests are stuck in unmanaged code not at a message pool but inside a third party library which is spending most of its time in `System.Globalization.CompareInfo.Compare(string, string, CompareOptions)`'s native implementation. But I'm stepping in and out in the managed debugger, so why no ThreadAbortException?

**Comment by jnm2** [comment]
Fun fact: If thread A calls `threadB.Abort`, thread A is blocked until threadB cleanly aborts. If thread C comes along and `TerminateThread`s thread B, sure enough it's nuked, but now thread A is blocked forever and the test run still doesn't complete.

**Comment by jnm2** [comment]
Probably the best thing for the test runner to do when a timeout happens is to fire and forget `new Thread(() => thread.Abort()) { Name = "Aborting test X" }.Start();` and return and report the failure even though the threads are still running. Then when all tests are done, the threads will go down with the process.

You could also consider passing a CancellationToken to test methods that have a parameter for it.

**Comment by jnm2** [comment]
Anyway @ChrisMaddock @CharliePoole (not sure which team to tag) I think that https://gist.github.com/jnm2/190bf111ef418901d7abb7542cd0a6be would be nice to merge into the test runner. What are your thoughts?

**Comment by oznetmaster** [comment]
@jnm2 Never seen that happen, and it should not:

> The Abort method is used to stop a managed thread permanently. When you call Abort, the common language runtime throws a ThreadAbortException in the target thread, which the target thread can catch. For more information, see Thread.Abort.
**System_CAPS_noteNote**
If a thread is executing unmanaged code when its Abort method is called, the runtime marks it ThreadState.AbortRequested. The exception is thrown when the thread returns to managed code.
Once a thread is aborted, it cannot be restarted.
The Abort method does not cause the thread to abort immediately, because the target thread can catch the ThreadAbortException and execute arbitrary amounts of code in a finally block. You can call Thread.Join if you need to wait until the thread has ended. Thread.Join is a blocking call that does not return until the thread has actually stopped executing or an optional timeout interval has elapsed. The aborted thread could call the ResetAbort method or perform unbounded processing in a finally block, so if you do not specify a timeout, the wait is not guaranteed to end.
Threads that are waiting on a call to the Thread.Join method can be interrupted by other threads that call Thread.Interrupt.

**Comment by jnm2** [comment]
> Never seen that happen, and it should not:

Me neither, but it's undeniable.

**Comment by oznetmaster** [comment]
Do you have a sample program (independent of NUnit) that exhibits this behavior?

**Comment by jnm2** [comment]
Shouldn't be hard, but you'll need to install a trial of DevExpress's Windows Forms library at a specific version that had a very bad perf bug. Or I'd have to spend a lot of time really understanding the phenomenon and build it from the ground up.

**Comment by oznetmaster** [comment]
Are you aborting a ThreadPool thread or a system thread?

If what you say is the case, then it should be easy to write a demonstration of it.

However, I have read both the .NET and the CLR code for Abort, and I do not see how it is possible.

**Comment by jnm2** [comment]
[Let's hop over here](https://gitter.im/nunit/nunit)

**Comment by oznetmaster** [comment]
On further research, this is apparently framework version dependent :(

On some versions, the Abort call does not return until the ThreadAbortException is thrown.  If the thread being aborted is currently running in native code, the exception is not thrown until control returns to managed code.  If that never happens, the calling thread will hang forever.

On other versions, the ThreadAborted flag is merely set, and the calling thread returns.

This means that there is no consistent way to deal with Thread.Abort.

**Comment by jnm2** [comment]
Here's more loveliness. I have a test that takes 3 seconds to run and it runs in 3 seconds every time... unless I set the timeout to 2 seconds (or otherwise cause a Thread.Abort 2 seconds in). Then it takes 35 to 40 seconds to run every time. It should be the exact same work both times, but not only is Abort not aborting, it's making it run more than ten times slower.

**Comment by jnm2** [comment]
I think I figured it out. There's a darn catch-all in DevExpress's code. So that's what's taking the abort 39 seconds to finish. It's in a hot path and which is repeatedly spending time in a catch-all. -_-
This is obscure so I wouldn't have NUnit worry about it.

**Comment by jnm2** [comment]
I think I figured it out. There's a darn catch-all in DevExpress's code. So that's what's taking the abort 39 seconds to finish. They're doing a quick sort and every compare operation goes to a comparer which wraps the comparison in in a catch-all. -_-
This is obscure so I wouldn't have NUnit worry about it.

Who even does that? You're not a message pump or thread pool, you have no business swallowing exceptions that don't belong to you. -_-

**Comment by jnm2** [comment]
Hmm, that was not charitable of me. I'll bet I do things that make even less sense. Anyway. This comes up a lot.

**Comment by jnm2** [comment]
@CharliePoole, /cc @ChrisMaddock 

> If you find any new wrinkles that work, consider adding them to ThreadUtility.Kill, which is used to terminate threads on timeout.

Very simply what works for me and what we could add is:

 * Abort like you're already doing
 * But before the abort, schedule a thread pool thread to check 100 ms later to see if the thread is aborted 
 * If the thread is indeed aborted, that's the end of that. But if it's not, the main thread will still be blocked inside the Thread.Abort call, so the threadpool thread posts a thread message to dislodge any possible native message pump and schedules another check 100 ms later.

Should I implement a PR?

**Comment by jnm2** [comment]
Hmm, looks like the master branch has a build error on TextOutputTests.cs, line 152.

**Comment by jnm2** [comment]
For reference, this is what it would look like: https://github.com/nunit/nunit/pull/2023
Supporting net20 made it a little noisy.

(I don't know what I did with ZenHub there.)

**Comment by CharliePoole** [comment]
@jnm2 Is this still an issue with the latest changes in how timeout works?

BTW - you should always triage an issue (label it) and assign yourself to it if you are working on it.

**Comment by jnm2** [comment]
@CharliePoole ugh, I'm so behind on everything. I'll try to get motion on this tomorrow.

**Comment by jnm2** [comment]
I can't do anything further here until https://github.com/nunit/docs/issues/163 produces some helpful result.

**Comment by CharliePoole** [comment]
@jnm2 Did you actually intend to close this?

**Comment by jnm2** [comment]
Yes, this was supposed to be closed in #2023.

**Comment by CharliePoole** [comment]
A comment is good in such situations.
